### DISCUSSION TITLE
Set ranking rules at search time

### DISCUSSION THREAD

**Discussion by loiclec** [meilisearch/product#593] [Feedback & Feature Proposal]
## Summary 
I would like to be able to set the ranking rules at search time, possibly with a new search parameter: `rankingRules`.

- If this parameter is not set (or = `null`), then the settings' ranking rules will be used
- If it is set, then the search query will be executed using the provided ranking rules

## Why

- There are no technical reasons for not being able to set the ranking rules at search time
- It would prove useful for users who want experiment with different ranking rules quickly and fine-tune the relevancy of their search results. Changing the ranking rules via a `settingsUpdate` task could take a very long time if a lot of indexing tasks are already enqueued. Doing it via a search parameter, on the other hand, is instant and does not modify the index.
- Quoting @gmourier: "[we could] customise the search experience PER user / user group thanks to tenant tokens (`rankingRules` would be a property of the tenant token, which would be enforced at search-time when it is used to authenticate the request)"

## Caveats

1. It could be dangerous to allow anyone to set arbitrary ranking rules due to performance (and therefore DoS) issues. This can be solved in two ways: (1) limit who can set custom ranking rules; or (2) make sure that searches stay quick no matter what ranking rules are used (my preferred solution)

2. Currently, one can add a custom ranking rule `my_field:asc` even if `my_field` is neither filterable nor sortable. This would not be possible at search time.

**Comment by tacman** [comment]
Here's an example of what I think is a pretty common use case.  I want to find Symfony bundles for working with APIs, so I set the search to "api" 

https://packages.survos.com/?q=api&marking=symfony_ok

![image](https://github.com/user-attachments/assets/92c554a4-afd4-42a2-9331-b41ab812dc29)

I'd like to look at the most widely-used ones first, so I sort by number of downloads.  As I'm scrolling down, I suddenly see the number of downloads changes, which jumps out as a bug.  But if you look at ranking, it's right.  So while I could reset the index to be prioritize sorting, most of the time I don't want to -- it's a runtime setting, not an index build setting.

![image](https://github.com/user-attachments/assets/b6e2b4c9-ba13-443d-ab1b-2823b1684a0e)

I think this is really common -- sort by price, rating, time,  combine with a full-text search.  Please consider making it a run-time option, even if there is a cost associated with pagination.

**Comment by tacman** [comment]
Here's an example on a single page:  https://packages.survos.com/?q=payment&marking=symfony_ok

![image](https://github.com/user-attachments/assets/038f9428-66cf-4fd9-a28d-23f7092f56f9)

Sorting by downloads (by clicking on the title) looks like it's broken.
